simulate/sam.go

```diff
@@ -53,53 +53,71 @@ func IlluminaPairedSam(refName string, ref []dna.Base, numPairs, readLen, avgFra
 
 // generateSamReadNoFlag generates a sam record for the input position.
 // Soft clips sequence that is off template and does not generate Flag, RNext, or PNext.
-func generateSamReadNoFlag(readName string, refName string, ref []dna.Base, start, end int) sam.Sam {
-	var s sam.Sam
-	s.QName = readName
-	s.Seq = make([]dna.Base, end-start)
+func generateSamReadNoFlag(readName string, refName string, ref []dna.Base, start, end int, flatErrorRate float64, alias numbers.BinomialAlias) sam.Sam {
+	var currSam sam.Sam
+	currSam.QName = readName
+	currSam.Seq = make([]dna.Base, end-start)
 	// generate qual
 	var bldr strings.Builder
-	for range s.Seq {
+	for range currSam.Seq {
 		bldr.WriteRune(rune(numbers.RandIntInRange(30, 40) + 33)) // high quality seq + ascii offset
 	}
-	s.Qual = bldr.String()
+	currSam.Qual = bldr.String()
 
 	// check if unmapped
 	if end < 0 || start > len(ref) {
-		s.RName = "*"
-		for i := range s.Seq {
-			s.Seq[i] = dna.Base(numbers.RandIntInRange(0, 4))
+		currSam.RName = "*"
+		for i := range currSam.Seq {
+			currSam.Seq[i] = dna.Base(numbers.RandIntInRange(0, 4))
 		}
-		return s
+		return currSam
 	}
 
-	s.MapQ = uint8(numbers.RandIntInRange(30, 40))
-	s.RName = refName
+	currSam.MapQ = uint8(numbers.RandIntInRange(30, 40))
+	currSam.RName = refName
 
 	// generate random seq if off template
 	var realSeqStartIdx, realSeqEndIdx int
 	for realSeqStartIdx = start; realSeqStartIdx < 0; realSeqStartIdx++ {
-		s.Seq[realSeqStartIdx-start] = dna.Base(numbers.RandIntInRange(0, 4))
+		currSam.Seq[realSeqStartIdx-start] = dna.Base(numbers.RandIntInRange(0, 4))
 	}
 	for realSeqEndIdx = end; realSeqEndIdx > len(ref); realSeqEndIdx-- {
-		s.Seq[len(s.Seq)-(1+(end-realSeqEndIdx))] = dna.Base(numbers.RandIntInRange(0, 4))
+		currSam.Seq[len(currSam.Seq)-(1+(end-realSeqEndIdx))] = dna.Base(numbers.RandIntInRange(0, 4))
+	}
+	copy(currSam.Seq[realSeqStartIdx-start:len(currSam.Seq)-(end-realSeqEndIdx)], ref[realSeqStartIdx:realSeqEndIdx])
+
+	// now we will simulate sequencing/PCR error with a flat error rate
+	if flatErrorRate > 0 {
+		numFlatErrors := numbers.RandBinomial(alias) // sample a binomial distribution to get the number of sequencing errors
```

Comment:
It seems like what is inside of this if-statement could be a function call where you pass a sam to have errors added to it and get it back